test/extended/pods/pdb.go

```diff
@@ -0,0 +1,182 @@
+				t.Kind = "Deployment"
+				meta = &t.ObjectMeta
+				selectorLabels = t.Spec.Selector
+				replicas = t.Status.Replicas
+			case *appsv1.StatefulSet:
+				t.Kind = "StatefulSet"
+				meta = &t.ObjectMeta
+				selectorLabels = t.Spec.Selector
+				replicas = t.Status.Replicas
+			case *appsv1.DaemonSet:
+				t.Kind = "DaemonSet"
+				meta = &t.ObjectMeta
+				selectorLabels = t.Spec.Selector
+				replicas = t.Status.DesiredNumberScheduled
+			default:
+				panic("not an object")
+			}
+			if !strings.Contains(meta.Namespace, "openshift-") {
+				// Not every component is following the convention of -operator at the end, there are some namespaces
+				// like openshift-monitoring and openshift-network-diagnostics and some of them run on worker nodes
+				// as well. So, need to figure out how to exclude operator components and components running on worker node
+				continue
+			}
+
+			key := fmt.Sprintf("%s/%s/%s", workload.GetObjectKind().GroupVersionKind().Kind, meta.Namespace, meta.Name)
+			var pdbs *v1beta1.PodDisruptionBudgetList
+			isPDBConfigured := false
+			var currentPDB v1beta1.PodDisruptionBudget
+			pdbs, err = kubeClient.PolicyV1beta1().PodDisruptionBudgets(meta.Namespace).List(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				e2e.Failf("Unexpected error getting PDBs: %v", err)
+			}
+			for _, pdb := range pdbs.Items {
+				if reflect.DeepEqual(pdb.Spec.Selector, selectorLabels) {
+					isPDBConfigured = true
+					currentPDB = pdb
+					operatorsHavingPDBS = append(operatorsHavingPDBS, key)
+					// The convention we came up with for static pod control plane components is to have a deployment
+					// doing health check for the corresponding component. This can be changed to labels later.
+					// As of now, I am assuming if the deployment name has `-guard` in the name and it runs in
+					// `openshift-*` namespace, it can be considered a deployment guarding control plane static pod
+					if strings.Contains(meta.Name, "-guard") &&
+						controlPlaneStaticPodNames.Has(meta.Name[:strings.LastIndex(meta.Name, "-")]) {
+						e2e.Logf("The workload %s is having a PDB %s to monitor the health of control plane component %s",
+							key, pdb.Name, meta.Name[:strings.LastIndex(meta.Name, "-")])
+					} else {
+						e2e.Logf("Workload %s has PDB %s configured", key, pdb.Name)
+					}
+					break
+				}
+			}
+			if inHAMode {
+				key = "HA/" + key
+				if !isPDBConfigured {
+					pdbMissingWorkloads = append(pdbMissingWorkloads, key)
+
+				} else if isPDBConfigured && replicas > 2 {
+					// We expect atleast 1 pod to be available while evicting/draining.
+					if currentPDB.Spec.MaxUnavailable != nil && currentPDB.Spec.MaxUnavailable.IntValue() < 1 {
+						pdbMisconfiguredWorkloads = append(pdbMisconfiguredWorkloads, key)
+					} else if currentPDB.Spec.MinAvailable != nil && currentPDB.Spec.MinAvailable.IntValue() < 1 {
+						pdbMisconfiguredWorkloads = append(pdbMisconfiguredWorkloads, key)
```

Comment:
We should probably add a message about why we are mad at this key, by either making `pdbMisconfiguredWorkloads` a map from workload key to string:

```go
pdbMisconfiguredWorkloads[key] = fmt.Sprintf("sets minAvailable %s, but $WHY_WE_WANT_IT_UNSET_OR_GREATER_THAN_ZERO", currentPDB.Spec.MinAvailable)
```

or a map from a misconfigured-reason to a slice of keys:

```go
pdbMisconfiguredWorkloads[pdbMaxUnavailableTooSmall] = append(pdbMisconfiguredWorkloads[pdbMaxUnavailableTooSmall], key)
```